### PR TITLE
Streamline webhook handler registration in `gardener-admission-controller`

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind cleanup technical-debt

**What this PR does / why we need it**:
This PR streamlines the webhook handler registration in `gardener-admission-controller` so that it's similar to what happens for the webhooks in `gardener-resource-manager` (see #6865), and generally similar to the registration of controllers in other components (`garden{er-controller-manager,er-scheduler,let}`).

**Which issue(s) this PR fixes**:
Part of #4251

**Special notes for your reviewer**:
Similar to #6865.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// AddToManager adds Handler to the given manager.
+func (h *Handler) AddToManager(ctx context.Context, mgr manager.Manager) error {
+	// Initialize caches here to ensure the readyz informer check will only succeed once informers required for this
+	// handler have synced so that http requests can be served quicker with pre-syncronized caches.
+	if _, err := mgr.GetCache().GetInformer(ctx, &corev1.Namespace{}); err != nil {
+		return err
+	}
+	if _, err := mgr.GetCache().GetInformer(ctx, &gardencorev1beta1.Project{}); err != nil {
+		return err
+	}
```
Comment (acumino) [inline pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go:43]: This was already there in the past, but why is this needed, will the manager will not have its cache synced already when its starts?

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go [hunk 2]
```diff
@@ -31,94 +34,78 @@ import (
...
-		requestLogger = requestLogger.WithValues("namespace", request.Namespace)
+// Handle checks the resource sizes.
+func (h *Handler) Handle(_ context.Context, req admission.Request) admission.Response {
+	var err error
+	switch req.Operation {
+	case admissionv1.Create:
+		err = h.handle(req)
+	case admissionv1.Update:
+		err = h.handle(req)
+	default:
```
Comment (acumino) [inline pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:57]: Here default case is admission errored, which means the deletion operation will give an error  `unknown operation request`. Is this expected?

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go [hunk 3]
```diff
@@ -31,94 +34,78 @@ import (
...
+		var apiStatus apierrors.APIStatus
+		if errors.As(err, &apiStatus) {
+			status := apiStatus.Status()
+			return admission.Response{AdmissionResponse: admissionv1.AdmissionResponse{Allowed: false, Result: &status}}
+		}
+		return admission.Denied(err.Error())
 	}
 
-	return response
+	return admission.Allowed("")
```
Comment (acumino) [inline pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:70]: Maybe we can write in the message `admission allowed` or something meaningful.

File: pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		)
+
+		metrics.RejectedResources.WithLabelValues(
+			string(req.Operation),
+			req.Kind.Kind,
+			req.Namespace,
+			metricReasonRejectedKubeconfig,
+		).Inc()
+
+		return fmt.Errorf("secret contains invalid kubeconfig: %w", err)
```
Comment (acumino) [inline pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go]: Previously errors were returned with some http responses messages also, now all errors will be plain without any classification of error in HTTP error message, we can use something similar to this - https://github.com/gardener/gardener/blob/584e63d5ae7672dd5f95f3af1fdfdcdd881465e0/pkg/seedadmissioncontroller/webhook/admission/extensionresources/admission.go#L115

File: pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go (hunk 1 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go:43]: Since this is a webhook server, it will only start the watches/informers when the first request is made. This is only happening when the handler receives the first request. This may take a while depending on the amount of resources.
To make sure the server is not ready without having initialized its caches, we get the informers already here when adding to make sure it syncs them on start.

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go (hunk 3 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:70]: This is not evaluated anyways, so there is no benefit in it. Also, it is not done in controller-runtime itself: https://github.com/gardener/gardener/blob/f88cb653da8bb153fa664912900f38eb0598540a/vendor/sigs.k8s.io/controller-runtime/pkg/webhook/admission/validator.go#L121

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go (hunk 2 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:57]: The deletion operation is not handled, so it seems to be reasonable to return an error instead of silently accepting it.

File: pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go (hunk 4 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go]: What would be the value of this? I don't think the errors are evaluated anywhere, are they? 👀 I would probably also drop this from GSAC when I merge it into GRM in a future PR.

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go (hunk 2 above)
Comment (acumino) [reply pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:57]: Maybe I am mistaken here,  but this webhook runs for every object, and if the object size is greater than the max limit it discards that object. So when an object is marked for deletion this webhook will be invoked and webhook will block the deletion of the object cause it returns an error instead of nil on the deletion request.

File: pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go (hunk 1 above)
Comment (acumino) [reply pkg/admissioncontroller/webhook/admission/namespacedeletion/add.go:43]: Thanks for the explanation. 😃

File: pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go (hunk 4 above)
Comment (acumino) [reply pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go]: I was not referring to this specific line but in general, for ex -  [ref](https://github.com/gardener/gardener/blob/f88cb653da8bb153fa664912900f38eb0598540a/pkg/admissioncontroller/webhook/admission/resourcesize/admission.go#L115), [ref](https://github.com/gardener/gardener/blob/f88cb653da8bb153fa664912900f38eb0598540a/pkg/admissioncontroller/webhook/admission/namespacedeletion/admission.go#L119) and similarly other places in other webhooks.

And these HTTP errors are parsed by the controller runtime here https://github.com/gardener/gardener/blob/f88cb653da8bb153fa664912900f38eb0598540a/vendor/sigs.k8s.io/controller-runtime/pkg/webhook/admission/validator.go#L112-L118.

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go (hunk 2 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:57]: The webhook only handles `CREATE` and `UPDATE` operations, so there should be no need to handle the `DELETE` operation: https://github.com/gardener/gardener/blob/33dc3fc1bbfc0f07b471e39f384431f0989a09eb/charts/gardener/controlplane/charts/application/templates/validatingwebhook-admission-controller.yaml#L73-L75

File: pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go (hunk 4 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/kubeconfigsecret/handler.go]: Done with 4a1f4ec696e9811f3aa1efcd02feda968d7002c1

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go [hunk 5]
```diff
@@ -31,94 +34,78 @@ import (
...
-	if !response.Allowed {
-		metrics.RejectedResources.WithLabelValues(
-			fmt.Sprint(request.Operation),
-			request.Kind.Kind,
-			request.Namespace,
-			metricReasonSizeExceeded,
-		).Inc()
+	if err != nil {
+		var apiStatus apierrors.APIStatus
+		if errors.As(err, &apiStatus) {
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:63]: Can you explain when this type of error is returned. IIUC, the `handle` function only returns custom `fmt.Errorf` errors. 👀

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		return fmt.Errorf("expected *corev1.Secret but got %T", obj)
+	}
+
+	alreadyExists, err := h.secretAlreadyExists(ctx, secret)
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return nil
+	}
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go]: Do we still need this check if down below in `internalDomainSecretExists` the search is restricted to the secret's namespace?

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+	oldSecret, ok := oldObj.(*corev1.Secret)
+	if !ok {
+		return fmt.Errorf("expected *corev1.Secret but got %T", oldObj)
+	}
+
+	seedName := gutil.ComputeSeedName(secret.Namespace)
+	if secret.Namespace != v1beta1constants.GardenNamespace && seedName == "" {
+		return nil
+	}
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go:90]: Why is this check only performed in the `update` case?

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	if exists {
+		return fmt.Errorf("cannot create internal domain secret because there can be only one secret with the 'internal-domain' secret role per namespace")
+	}
+
+	_, _, _, _, _, err = gutil.GetDomainInfoFromAnnotations(secret.Annotations)
+	return err
+}
+
+// ValidateUpdate performs the check.
+func (h *Handler) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go:73]: The `GardenRoleInternalDomain` label can also be added later on, through an update. Shouldn't we perform the uniqueness of the secret here as well?

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice to see some boilerplate code being eliminated ⭐
Please see my questions and comments added inline.

File: pkg/admissioncontroller/webhook/admission/resourcesize/handler.go (hunk 5 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/resourcesize/handler.go:63]: After https://github.com/gardener/gardener/pull/6873/commits/4a1f4ec696e9811f3aa1efcd02feda968d7002c1, it can also return `APIStatus` errors as requested by https://github.com/gardener/gardener/pull/6873#discussion_r1007786239.

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go (hunk 6 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go]: Probably not, I remove it.

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go (hunk 7 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go:90]: It was done only in `update` and `delete`, but it should also have been done in `create`. Good catch, I'll add it there as well.

File: pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go (hunk 8 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/internaldomainsecret/handler.go:73]: OK, done

**Review by timuthy — COMMENTED** [review]
Thanks for addressing the comments.
/lgtm

**Review by acumino — APPROVED** [review]
/lgtm
